### PR TITLE
Add new fuzzer for oss-fuzz

### PR DESCRIPTION
**Pull request by DavidKorczynski**
Hi,

I have been working on extending the fuzzing of libsdnfile and would be very happy to start contributing more. This PR adds an initial fuzzer targeting `sf_read_double`, `sf_read_int` and `sf_read_short`. I think there are a few more things we could do to improve the fuzzing of libsndfile - let me know if you would appreciate more contributions in this area. 

David

### CODE REVIEW

**Comment by evpobr** [comment]
Hi, @DavidKorczynski , thanks for pull request!

Let's ask the author of original fuzzer first. @cmeister2 , can you review?

**Review by evpobr — CHANGES_REQUESTED** [review]
Could you reformat code using our coding style?

**Comment by cmeister2** [comment]
FYI I will take a look at this on Monday when I'm back at a computer, if
that's ok with you.

File: ossfuzz/sndfile_alt_fuzzer.cc [hunk 1]
```diff
@@ -0,0 +1,200 @@
+#include <stdint.h>
```
Comment (SoapGentoo) [inline ossfuzz/sndfile_alt_fuzzer.cc:1]: not needed with `<inttypes.h>`

File: ossfuzz/sndfile_alt_fuzzer.cc [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size)
+{
+
+  if (size == 0) {
+     return 0;
+  }
+  const uint8_t decider = *data;
+  data++; 
+  size--;
```
Comment (SoapGentoo) [inline ossfuzz/sndfile_alt_fuzzer.cc:105]: prefix operators over postfix operators

File: ossfuzz/sndfile_alt_fuzzer.cc [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+EXIT_LABEL:
+
+    if (sndfile != NULL)
+    {
+      sf_close(sndfile);
+    }
+
+
```
Comment (SoapGentoo) [inline ossfuzz/sndfile_alt_fuzzer.cc:198]: unnecessary whitespace

**Review by SoapGentoo — CHANGES_REQUESTED** [review]

File: ossfuzz/sndfile_alt_fuzzer.cc (hunk 1 above)
Comment (DavidKorczynski) [reply ossfuzz/sndfile_alt_fuzzer.cc]: Thanks!

File: ossfuzz/sndfile_alt_fuzzer.cc (hunk 2 above)
Comment (DavidKorczynski) [reply ossfuzz/sndfile_alt_fuzzer.cc]: I switched to `data += 1;` and `size -= 1;` to avoid the inc/dec. Does that work too?

File: ossfuzz/sndfile_alt_fuzzer.cc (hunk 3 above)
Comment (DavidKorczynski) [reply ossfuzz/sndfile_alt_fuzzer.cc:200]: Thanks!

**Comment by DavidKorczynski** [comment]
> FYI I will take a look at this on Monday when I'm back at a computer, if that's ok with you.
> […](#)

Yes, thanks!

**Comment by DavidKorczynski** [comment]
> Could you reformat code using our coding style?

Yes

I followed the style I observed in the existing fuzzer https://github.com/libsndfile/libsndfile/blob/master/ossfuzz/sndfile_fuzzer.cc

But it seems that it doesn't follow the code style of the rest of the repository, is that correctly understood? Is there a canonical place to find code style documentation?

File: ossfuzz/sndfile_alt_fuzzer.cc (hunk 2 above)
Comment (SoapGentoo) [reply ossfuzz/sndfile_alt_fuzzer.cc]: thats fine, but what was the issue with the increment?

**Comment by evpobr** [comment]
> > Could you reformat code using our coding style?
> 
> Yes
> 
> I followed the style I observed in the existing fuzzer https://github.com/libsndfile/libsndfile/blob/master/ossfuzz/sndfile_fuzzer.cc
> 
> But it seems that it doesn't follow the code style of the rest of the repository, is that correctly understood? Is there a canonical place to find code style documentation?

In general, we try to follow the strange style of code that has developed here historically. There are no rules described, all that remains is to study the code.

**Comment by DavidKorczynski** [comment]
> In general, we try to follow the strange style of code that has developed here historically. There are no rules described, all that remains is to study the code.

Thanks! But there are some parts of the code that look different, for example the switch cases here:

https://github.com/libsndfile/libsndfile/blob/dc1c8c2492a776c60360fa0054978ad0352aace6/src/avr.c#L128-L147

is very different to the style of the switch in the fuzzer:

https://github.com/libsndfile/libsndfile/blob/dc1c8c2492a776c60360fa0054978ad0352aace6/ossfuzz/sndfile_fuzzer.cc#L26-L42

I went with a similar style to the one in the fuzzer so unless I got that style wrong then I am not sure what to change to make the code follow your code style.

**Comment by evpobr** [comment]
The first one is (unfortunately) canonical.

File: ossfuzz/sndfile_alt_fuzzer.cc [hunk 4]
```diff
@@ -0,0 +1,202 @@
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sndfile.h>
+#include <inttypes.h>
+
+typedef struct
```
Comment (cmeister2) [inline ossfuzz/sndfile_alt_fuzzer.cc:7]: General comment: a lot of this code looks copied and pasted. Could you possibly split out the common elements from both this and the standard fuzzer and have them in common headers and common source files? Things like VIO_DATA and all the vf* functionality would be a good place to start.

Comment (DavidKorczynski) [reply ossfuzz/sndfile_alt_fuzzer.cc]: I fixed this with a `sndfile_fuzz_header.h` file. What do you think?

**Review by evpobr — APPROVED** [review]

**Comment by evpobr** [comment]
@cmeister2 , @SoapGentoo , is it OK?

File: ossfuzz/sndfile_alt_fuzzer.cc [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+#include "sndfile_fuzz_header.h"
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size)
+{   // One byte is needed for deciding which function to target.
+    if (size == 0)
+       return 0 ;
+
+    const uint8_t decider = *data ;
+    data += 1 ;
+    size -= 1 ;
```
Comment (SoapGentoo) [inline ossfuzz/sndfile_alt_fuzzer.cc:16]: I would have still preferred
```
++data ;
--size ;
```
but whatever

**Review by SoapGentoo — CHANGES_REQUESTED** [review]

**Review by SoapGentoo — APPROVED** [review]

File: ossfuzz/sndfile_fuzz_header.h [hunk 6]
```diff
@@ -0,0 +1,114 @@
+typedef struct
```
Comment (cmeister2) [inline ossfuzz/sndfile_fuzz_header.h:1]: I'd have expected to see an ifdef guard here, e.g. https://github.com/libsndfile/libsndfile/blob/master/src/mpeg.h#L20

```
#ifndef SNDFILE_FUZZ_HEADER_H
#define SNDFILE_FUZZ_HEADER_H
```
with the closing `#endif` at the end.

**Comment by cmeister2** [comment]
One minor comment but otherwise good.

**Comment by evpobr** [comment]
I have polished this PR a little, if no one will mind, I will squash and merge it tonight.

**Comment by DavidKorczynski** [comment]
> I have polished this PR a little, if no one will mind, I will squash and merge it tonight.

sounds good to me! Thanks!

**Comment by evpobr** [comment]
@DavidKorczynski , next time, [create a new branch for the pull request](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests). It's not very cool that I can push into your master branch.

**Comment by DavidKorczynski** [comment]
Thanks for the reviews and thanks for merging in @evpobr - and thanks for the nit on the master branch.
